pkg/gitprovider/azure-devops.go

```diff
@@ -0,0 +1,483 @@
+		return nil, errors.New("cannot parse git URL: " + gitURL)
+	}
+
+	repo.Source = u.Host
+	repo.Owner = matches[3]
+	repo.Name = matches[5]
+	projectName := matches[4]
+
+	urlPath := strings.TrimPrefix(u.Path, "/")
+
+	if projectName == "" {
+		projectName = repo.Name
+		parts := strings.SplitN(urlPath, fmt.Sprintf("%s/", repo.Owner), 2)
+		if len(parts) != 2 {
+			return nil, errors.New("cannot parse git URL: " + gitURL)
+		}
+
+		urlPath = strings.Join([]string{parts[0], repo.Owner, projectName, parts[1]}, "/")
+		urlPath = strings.TrimPrefix(urlPath, "/")
+	}
+
+	parts := strings.Split(urlPath, "/")
+
+	repo.Url = g.getAzureDevopsCloneUrl(repo.Source, repo.Owner, repo.Name, projectName)
+	repo.Name, _ = url.QueryUnescape(repo.Name)
+	projectName, _ = url.QueryUnescape(projectName)
+
+	repoId, err := g.getAzureDevopsRepoId(repo.Name, projectName)
+	if err != nil {
+		return nil, errors.New("cannot parse git URL: " + gitURL)
+	}
+	repo.Id = repoId
+
+	queryParams, err := url.QueryUnescape(u.RawQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(parts) <= 4 && queryParams == "" {
+		return repo, nil
+	}
+
+	switch {
+	case len(parts) >= 6 && parts[4] == "pullrequest":
+		prNumber, _ := strconv.Atoi(parts[5])
+		prUint := uint32(prNumber)
+		repo.PrNumber = &prUint
+		repo.Path = nil
+	case len(parts) >= 6 && parts[4] == "commit":
+		repo.Sha = &parts[5]
+		repo.Branch = &parts[5]
+		repo.Path = nil
+	}
+
+	switch {
+	case strings.Contains(queryParams, "itemVersion=GB"):
+		fallthrough
+	case strings.Contains(queryParams, "version=GB"):
+		fallthrough
+	case strings.Contains(queryParams, "refName=refs/heads/"):
+		re := regexp.MustCompile(`(itemVersion|version|refName)=(GB|GT|refs/heads/)(.+?)(&|$)`)
+		matches := re.FindStringSubmatch(queryParams)
+		if len(matches) != 5 {
+			return nil, errors.New("cannot parse git URL: " + gitURL)
+		}
+
+		repo.Branch = &matches[3]
+		repo.Path = nil
+	case strings.Contains(queryParams, "path="):
+		re := regexp.MustCompile(`path=/(.+?)(&|$)`)
+		matches := re.FindStringSubmatch(queryParams)
+		if len(matches) != 3 {
+			return nil, errors.New("cannot parse git URL: " + gitURL)
+		}
+
+		repo.Path = &matches[1]
+	}
+
+	return repo, nil
+}
+
+func (g *AzureDevOpsGitProvider) getAzureDevopsCloneUrl(source string, owner string, repo string, project string) string {
+	return fmt.Sprintf("https://%s/%s/%s/_git/%s", source, owner, project, repo)
+}
+
+func (g *AzureDevOpsGitProvider) getAzureDevopsRepoId(repo string, project string) (string, error) {
+	client, err := g.getGitClient()
+	if err != nil {
+		return "", err
+	}
+
+	repository, err := client.GetRepository(context.Background(), git.GetRepositoryArgs{
+		RepositoryId: &repo,
+		Project:      &project,
+	})
+	if err != nil {
+		return "", err
+	}
+
+	return repository.Id.String(), nil
+}
+
+func (g *AzureDevOpsGitProvider) getOwnerName() string {
+	baseUrl := g.baseApiUrl
+	re := regexp.MustCompile(`https://dev.azure.com/(.+)`)
```

Comment:
Why is the URL hardcoded here and not inferred from `g.baseApiUrl`?

Reply:
done